### PR TITLE
Improvement/background sync

### PR DESCRIPTION
**Pull request by neddoz**
# what this PR does?
* Fixes: https://github.com/raedahgroup/dcrios/issues/409

### CODE REVIEW

File: Decred Wallet/Features/Wallet Utils/Syncer.swift [hunk 1]
```diff
@@ -188,6 +196,19 @@ class Syncer: NSObject, AppLifeCycleDelegate {
         }
     }
 
+    func applicationWillTerminate() {
+        if backgroundTask != .invalid {
+            endBackgroundTask()
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Wallet Utils/Syncer.swift]: Add a print statement here to print `app terminated`

File: Decred Wallet/Features/Wallet Utils/Syncer.swift [hunk 2]
```diff
@@ -208,6 +229,29 @@ class Syncer: NSObject, AppLifeCycleDelegate {
...
+
+    @objc private func reinstateBackgroundTask() {
+        if backgroundTask ==  .invalid {
+            registerBackgroundTask()
+        }
+    }
+
+    private func registerBackgroundTask() {
+        backgroundTask = UIApplication.shared.beginBackgroundTask { [weak self] in
+            self?.endBackgroundTask()
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Wallet Utils/Syncer.swift]: Add a print statement here: `background task expired`

File: Decred Wallet/Features/Wallet Utils/Syncer.swift [hunk 3]
```diff
@@ -208,6 +229,29 @@ class Syncer: NSObject, AppLifeCycleDelegate {
             self.networkLastActive = nil // Network is active at this point.
         }
     }
+
+    @objc private func reinstateBackgroundTask() {
+        if backgroundTask ==  .invalid {
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Wallet Utils/Syncer.swift]: Should also check `if currentSyncOp != .Done`

File: Decred Wallet/Features/Wallet Utils/Syncer.swift [hunk 4]
```diff
@@ -208,6 +229,29 @@ class Syncer: NSObject, AppLifeCycleDelegate {
...
+
+    private func registerBackgroundTask() {
+        backgroundTask = UIApplication.shared.beginBackgroundTask { [weak self] in
+            self?.endBackgroundTask()
+        }
+        assert(backgroundTask != .invalid)
+    }
+
+    private func endBackgroundTask() {
+        print("Background task ended at: ", Date())
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Wallet Utils/Syncer.swift]: This seems to be called multiple times, not really sure why. May be a good idea to check that `backgroundTask != .invalid` before proceeding here.

File: Decred Wallet/Features/Wallet Utils/Syncer.swift (hunk 2 above)
Comment (neddoz) [reply Decred Wallet/Features/Wallet Utils/Syncer.swift]: This is already handled in the `endBackgroundTask()` method where there is a print statement indicated the time it expired.

File: Decred Wallet/Features/Wallet Utils/Syncer.swift (hunk 4 above)
Comment (neddoz) [reply Decred Wallet/Features/Wallet Utils/Syncer.swift]: > Just noticed that sync continues even after the background task is ended. See the log output here https://pastebin.com/Aqkkxp9g.
> 
> Also makes me wonder, what's the behaviour when the background task expires and is ended by the OS?

Should be fine now. Also, notice for every time the app goes into background we deregister any previous background task and register a new one.

File: Decred Wallet/AppDelegate.swift [hunk 5]
```diff
@@ -20,6 +20,8 @@ import Crashlytics
 protocol AppLifeCycleDelegate {
     func applicationEnteredForegroundFromSuspendedState(_ lastActiveTime: Date)
     func networkChanged(_ connection: Reachability.Connection)
+    func applicationWillTerminate()
+    func applicationWillEnterBackground()
```
Comment (itswisdomagain) [inline Decred Wallet/AppDelegate.swift]: Re-order these methods in the following order
`applicationEnteredForegroundFromSuspendedState`
`applicationWillEnterBackground`
`applicationWillTerminate`
`networkChanged`

The implementations in `Syncer.swift` should also follow same order.

File: Decred Wallet/Features/Wallet Utils/Syncer.swift [hunk 6]
```diff
@@ -208,6 +227,22 @@ class Syncer: NSObject, AppLifeCycleDelegate {
             self.networkLastActive = nil // Network is active at this point.
         }
     }
+
+    private func registerBackgroundTask() {
+        backgroundTask = UIApplication.shared.beginBackgroundTask { [weak self] in
+            self?.endBackgroundTask()
```
Comment (itswisdomagain) [inline Decred Wallet/Features/Wallet Utils/Syncer.swift]: Add a print statement before this line 233: `print("Background task expired")`

**Pull request merged**